python/comps.py

```diff
@@ -17,213 +16,71 @@ def get_comps(
-    ).apply(
-        lambda x: x.index.tolist()
-    ).apply(
-        lambda x: [x[0], x[-1]]
-    )
-    bin_argmin = [idx[0] for idx in price_bin_indices.values]
-    bin_argmax = [idx[1] for idx in price_bin_indices.values]
-    price_bin_indices = pd.DataFrame(
-        {
-            "id": price_bin_indices.index.astype(int),
-            "argmin": bin_argmin,
-            "argmax": bin_argmax,
-            "min": sorted_comparison_df["predicted_value"][bin_argmin].values,
-            "max": sorted_comparison_df["predicted_value"][bin_argmax].values
-        },
-    )
-
-    # Fudge the minimum/maximum values in the price bin range so that they
-    # approximate infinity. This is necessary because observation data may
-    # have observed values that are lower than the minimum value in the
-    # comparison data, or higher than the maximum value
-    price_bin_indices.at[0, "min"] = np.iinfo(np.int32).min + 1
-    price_bin_indices.at[num_price_bins - 1, "max"] = np.iinfo(np.int32).max - 1
-
-    # Update bins to ensure they have no gaps. This enables us to put values
-    # from the observation dataframe into these bins, since otherwise
-    # values in the observation dataframe might be in between the maximum
-    # and minimum values of two adjacent bins
-    for bin_idx in range(len(price_bin_indices)):
-        # Skip the first bin, since its min val should aways be the smallest
-        # possible int
-        if bin_idx > 0:
-            price_bin_indices.at[bin_idx, "min"] = (
-                price_bin_indices["max"][bin_idx - 1] + 1
-            )
-
-    # Place observations in bins. Do this in a numba-accelerated function so
-    # that we can make use of fast loops
-    observation_df["price_bin"] = _bin_by_price(
-        observation_df[["id", "predicted_value"]].values,
-        price_bin_indices.values
+    # Chunk the observations so that the script can periodically report progress
+    num_chunks = 10
+    observation_df["chunk"] = pd.cut(
+      observation_df.index,
+      bins=num_chunks,
+      labels=False
     )
 
     total_num_observations = len(observation_df)
-    total_num_possible_comps = len(sorted_comparison_df)
-    binned_ids, binned_scores = [], []
-    for bin_idx, bin in price_bin_indices.iterrows():
-        observations = observation_df[observation_df["price_bin"] == bin["id"]]
-
-        if observations.empty:
-            print(
-                f"No observations in bin {bin['id'] + 1}; skipping",
-                # Flush statement to stdout so that reticulate will print it
-                # in real time
-                flush=True
-            )
-            continue
-
-        observation_matrix = observations.drop(
-          columns=["id", "predicted_value", "price_bin"]
-        ).values
+    total_num_possible_comps = len(comparison_df)
+    chunked_ids, chunked_scores = [], []
+    for chunk_num in set(observation_df["chunk"]):
```

Comment:
We iterate over the set of chunks here rather than something like `range(1, num_chunks + 1)` because `pd.cut` won't produce a bin for each number in the range if the number of targets is less than the number of bins. This should never occur in production, but it can happen if you're testing the algorithm on a very small number of targets.